src/System.Management.Automation/engine/Attributes.cs

```diff
@@ -1891,20 +1920,23 @@ public sealed class ValidateNotNullAttribute : ValidateArgumentsAttribute
         /// </exception>
         protected override void Validate(object arguments, EngineIntrinsics engineIntrinsics)
         {
-            IEnumerable ienum = null;
-            IEnumerator itor = null;
-
             if (arguments == null || arguments == AutomationNull.Value)
             {
                 throw new ValidationMetadataException(
                     "ArgumentIsNull",
                     null,
                     Metadata.ValidateNotNullFailure);
             }
-            else if ((ienum = arguments as IEnumerable) != null)
+            else if (IsArgumentCollection(arguments.GetType(), out bool isElementValueType))
             {
-                foreach (object element in ienum)
+                // If the element of the collection is of value type, then no need to check for null
+                // because a value-type value cannot be null.
+                if (isElementValueType) { return; }
+
+                IEnumerator ienum = LanguagePrimitives.GetEnumerator(arguments);
+                while (ienum.MoveNext())
```

Comment:
Should we really check that every element is not null?  To me, the fact that you have at least one element should satisfy the condition?